File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */

var lowestCommonAncestor = function(root, p, q) {

    let currentNode = root;
    
    while (currentNode) {
        const valP = p.val;
        const valQ = q.val;
        const valCur = currentNode.val;
        
        if (valP > valCur && valQ > valCur) {
            currentNode = currentNode.right;
        } else if (valP < valCur && valQ < valCur) {
            currentNode = currentNode.left;
        } else {
            return currentNode;
        }
    } 
};

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {TreeNode} p
# @param {TreeNode} q
# @return {TreeNode}
def lowest_common_ancestor(root, p, q)
    p_val = p.val
    q_val = q.val
    root_val = root.val
   return root if (p_val - root_val) *  (q_val - root_val) <= 0
    
   if q_val < root_val
       lowest_common_ancestor(root.left, p, q)
   else 
       lowest_common_ancestor(root.right, p, q)
   end
end